packages/theme/src/cli/utilities/theme-environment/html.ts

```diff
@@ -98,3 +102,33 @@ function getErrorPage(options: {title: string; header: string; message: string;
     </body>
   </html>`
 }
+
+function assertThemeId(response: Response, html: string, expectedThemeId: string) {
+  /**
+   * DOM example:
+   *
+   * ```
+   * <script>var Shopify = Shopify || {};
+   * Shopify.locale = "en";
+   * Shopify.theme = {"name":"Development","id":143509762348,"theme_store_id":null,"role":"development"};
+   * Shopify.theme.handle = "null";
+   * ...;</script>
+   * ```
+   */
+  const obtainedThemeId = html.match(/Shopify\.theme\s*=\s*{[^}]+?"id":\s*"?(\d+)"?(}|,)/)?.[1]
```

Comment:
One alternative way of implementing this would be relying on the JavaScript runtime.

So, we could perform this logic at `packages/theme/src/cli/utilities/theme-environment/hot-reload/client.ts` and the client would ping the server, letting it know that the process should be terminated because the theme ID is different.

I slightly prefer the JS runtime approach because we [support](https://shopify.dev/docs/api/ajax/reference/cart) the `Shopify` global object, while the shape of the declaration might change.

Still, the current approach works as expected, and the shape of the declaration is unlikely to change, so feel free to keep it as is if you prefer this approach more, @frandiox. One benefit of the current approach is that it's more self-contained.